hpx::util::optional fails

## Expected Behavior

The following piece of code should compile:
```cpp
#include <hpx/util/optional.hpp>
#include <hpx/hpx_main.hpp>

int main() {
    hpx::util::optional<int> x;
    int y = 1;
    x = y;
}
```
## Actual Behavior

The following error messages are produced:
```
------ Build started: Project: phylanx_component, Configuration: Debug x64 ------
program.cpp
hpx\util\optional.hpp(254): error C2660: 'hpx::util::optional<int>::swap': function does not take 2 arguments
hpx\util\optional.hpp(151): note: while compiling class template member function 'hpx::util::optional<int> &hpx::util::optional<int>::operator =(const T &)'
        with
        [
            T=int
        ]
program.cpp(229): note: see reference to function template instantiation 'hpx::util::optional<int> &hpx::util::optional<int>::operator =(const T &)' being compiled
        with
        [
            T=int
        ]
program.cpp(227): note: see reference to class template instantiation 'hpx::util::optional<int>' being compiled
hpx\parallel\execution_policy.hpp(675): note: see reference to class template instantiation 'hpx::parallel::execution::parallel_policy_executor<hpx::launch>' being compiled
hpx\util\optional.hpp(254): error C2056: illegal expression
Done building project "phylanx_component.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 1 up-to-date, 0 skipped ==========
```

For the record, the following code works:
```cpp
#include <hpx/util/optional.hpp>
#include <hpx/hpx_main.hpp>

int main() {
    hpx::util::optional<int> x;
    int y = 1;
    x = std::move(y);
}
```

--- a/hpx/util/optional.hpp
+++ b/hpx/util/optional.hpp
@@ -42,6 +42,15 @@ namespace hpx { namespace util
         }
     };
 
+    namespace _optional_swap
+    {
+        using std::swap;
+
+        template <typename T>
+        void check_swap() noexcept(noexcept(
+            swap(std::declval<T&>(), std::declval<T&>())));
+    }
+
     template <typename T>
     class optional
     {
@@ -249,9 +258,9 @@ namespace hpx { namespace util
             empty_ = false;
         }
 
-        void swap(optional& other)
-            noexcept(std::is_nothrow_move_constructible<T>::value &&
-                noexcept(swap(std::declval<T&>(), std::declval<T&>())))
+        void swap(optional& other) noexcept(
+            std::is_nothrow_move_constructible<T>::value &&
+            noexcept(_optional_swap::check_swap<T>()))
         {
             // do nothing if both are empty
             if (empty_ && other.empty_)
@@ -261,6 +270,7 @@ namespace hpx { namespace util
             // swap content if both are non-empty
             if (!empty_ && !other.empty_)
             {
+                using std::swap;
                 swap(**this, *other);
                 return;
             }

--- a/tests/regressions/util/CMakeLists.txt
+++ b/tests/regressions/util/CMakeLists.txt
@@ -11,6 +11,7 @@ set(tests
     function_argument
     function_serialization_728
     iarchive_1237
+    optional_swap_3200
     protect_with_nullary_pfo
     serialize_buffer_1069
     set_config_entry_deadlock

--- a/tests/regressions/util/optional_swap_3200.cpp
+++ b/tests/regressions/util/optional_swap_3200.cpp
@@ -1,0 +1,20 @@
+//  Copyright (c) 2018 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/hpx_main.hpp>
+#include <hpx/util/optional.hpp>
+#include <hpx/util/lightweight_test.hpp>
+
+int main()
+{
+    hpx::util::optional<int> x;
+    int y = 42;
+    x = y;
+
+    HPX_TEST(x.has_value());
+    HPX_TEST_EQ(*x, 42);
+
+    return hpx::util::report_errors();
+}

